File: CTCI/Python/sorting.py
def bubble_sort(array):
    i = 0
    while i < len(array):
        j = i + 1
        while j < len(array):
            if array[i] > array[j]:
                array[i], array[j] = array[j], array[i]
            j += 1
        i += 1

    return array

print bubble_sort([1,2,45,1,3,4,7,2,1,100,312])

def merge_sort(array):
    if len(array) < 2:
        return array
    mid = len(array) / 2
    left = merge_sort(array[0:mid])
    right = merge_sort(array[mid:])

    return merge(left, right)

def merge(left, right):
    result = []
    i, j = 0, 0

    while len(left) > j and len(right) > i:
        if left[j] > right[i]:
            result.append(right[i])
            i += 1
        else:
            result.append(left[j])
            j += 1

    while j < len(left):
        result.append(left[j])
        j += 1
    while i < len(right):
        result.append(right[i])
        i += 1

    return result

print merge_sort([1,2,45,1,3,4,7,2,1,100,312])

File: CTCI/JavaScript/sorting.js
const bubbleSort = (array) => {
  for (let i = 0; i < array.length; i++) {
    for (let j = i + 1; j < array.length; j++) {
      if (array[i] > array[j]) {
        let temp = array[i];
        array[i] = array[j];
        array[j] = temp;
      }
    }
  }

  return array;
}

console.log(bubbleSort([1,2,45,1,3,4,7,2,1,100,312]));

const mergeSort = (array) => {
  let len = array.length;

  if (array.length < 2) return array;

  let mid = Math.floor(len / 2);
  let left = mergeSort(array.slice(0, mid));
  let right = mergeSort(array.slice(mid));

  return merge(left, right);
}

const merge = (left, right) => {
  let result = [];

  while (left.length > 0 && right.length > 0) {
    if (left[0] > right[0]) {
      result.push(right.shift());
    } else {
      result.push(left.shift());
    }
  }

  return result.concat(left, right);
}

console.log(mergeSort([1,2,45,1,3,4,7,2,1,100,312]));
